evm_arithmetization/src/cpu/kernel/interpreter.rs

```diff
@@ -23,50 +24,39 @@ use crate::extension_tower::BN_BASE;
 use crate::generation::mpt::load_all_mpts;
 use crate::generation::prover_input::ProverInputFn;
 use crate::generation::rlp::all_rlp_prover_inputs_reversed;
-use crate::generation::state::{all_withdrawals_prover_inputs_reversed, GenerationState};
-use crate::generation::GenerationInputs;
+use crate::generation::state::{
+    self, all_withdrawals_prover_inputs_reversed, GenerationState, GenerationStateCheckpoint,
+};
+use crate::generation::{run_cpu, GenerationInputs, State};
 use crate::memory::segments::{Segment, SEGMENT_SCALING_FACTOR};
 use crate::util::{h2u, u256_to_u8, u256_to_usize};
 use crate::witness::errors::{ProgramError, ProverInputError};
 use crate::witness::gas::gas_to_charge;
-use crate::witness::memory::{MemoryAddress, MemoryContextState, MemorySegmentState, MemoryState};
+use crate::witness::memory::{
+    MemoryAddress, MemoryContextState, MemoryOp, MemoryOpKind, MemorySegmentState, MemoryState,
+};
 use crate::witness::operation::{Operation, CONTEXT_SCALING_FACTOR};
 use crate::witness::state::RegistersState;
-use crate::witness::transition::decode;
-use crate::witness::util::stack_peek;
+use crate::witness::util::{push_no_write, stack_peek};
+use crate::{arithmetic, logic};
+
+type F = GoldilocksField;
 
 /// Halt interpreter execution whenever a jump to this offset is done.
 const DEFAULT_HALT_OFFSET: usize = 0xdeadbeef;
 
-impl MemoryState {
-    pub(crate) fn mload_general(&self, context: usize, segment: Segment, offset: usize) -> U256 {
-        self.get(MemoryAddress::new(context, segment, offset))
-    }
-
-    fn mstore_general(
-        &mut self,
-        context: usize,
-        segment: Segment,
-        offset: usize,
-        value: U256,
-    ) -> InterpreterMemOpKind {
-        let old_value = self.mload_general(context, segment, offset);
-        self.set(MemoryAddress::new(context, segment, offset), value);
-        InterpreterMemOpKind::Write(old_value, context, segment as usize, offset)
-    }
-}
-
-pub(crate) struct Interpreter<'a, F: Field> {
+pub(crate) struct Interpreter<F: Field> {
     pub(crate) generation_state: GenerationState<F>,
-    prover_inputs_map: &'a HashMap<usize, ProverInputFn>,
     pub(crate) halt_offsets: Vec<usize>,
     // The interpreter will halt only if the current context matches halt_context
-    halt_context: Option<usize>,
+    pub(crate) halt_context: Option<usize>,
     pub(crate) debug_offsets: Vec<usize>,
-    running: bool,
-    opcode_count: [usize; 0x100],
+    pub(crate) opcode_count: [usize; 0x100],
     memops: Vec<InterpreterMemOpKind>,
     jumpdest_table: HashMap<usize, BTreeSet<usize>>,
+    pub(crate) preinitialized_segments: HashMap<Segment, MemorySegmentState>,
+    pub(crate) is_jumpdest_analysis: bool,
+    pub(crate) clock: usize,
 }
```

Comment:
Docs for these members. The `Interpreter` type is key to understand the rest of the code, so I think it deserves some docs even though the members are only `pub(crate)`.